sam_view.c

```diff
@@ -560,211 +560,425 @@ static const char *copied_tags[] = { "RG", "BC", "QT", NULL };
 
 static void bam2fq_usage(FILE *to)
 {
-    fprintf(to, "\nUsage:   samtools bam2fq [-nOt] [-s <outSE.fq>] <in.bam>\n\n");
-    fprintf(to, "Options: -n        don't append /1 and /2 to the read name\n");
-    fprintf(to, "         -O        output quality in the OQ tag if present\n");
-    fprintf(to, "         -s FILE   write singleton reads to FILE [assume single-end]\n");
-    fprintf(to, "         -t        copy RG, BC and QT tags to the FASTQ header line\n");
-    fprintf(to, "\n");
+    fprintf(to,
+"Usage:   samtools bam2fq [-nOt] [-s <outSE.fq>] <in.bam>\n\n"
+"Options: -n        don't append /1 and /2 to the read name\n"
+"         -O        output quality in the OQ tag if present\n"
+"         -v INT    default quality score if not given in file [1]\n"
+"         -s FILE   write singleton reads to FILE [assume single-end]\n"
+"         -t        copy RG, BC and QT tags to the FASTQ header line\n"
+"         -1 FILE   write paired reads flagged READ1 to FILE\n"
+"         -2 FILE   write paired reads flagged READ2 to FILE\n"
+"         -0 FILE   write paired reads flagged both or neither READ1 and READ2 to FILE\n"
+"         -f INT    only include reads with all bits set in INT set in FLAG [0]\n"
+"         -F INT    only include reads with none of the bits set in INT\n"
+"                   set in FLAG [0]\n"
+"         -T TYPE   write the output in FASTQ or FASTA [FASTQ]\n");
 }
 
-int main_bam2fq(int argc, char *argv[])
-{
+typedef enum { READ_UNKNOWN = 0, READ_1 = 1, READ_2 = 2 } readpart;
+typedef enum { FASTA, FASTQ } fastfile;
+typedef struct bam2fq_opts {
+    char *fnse;
+    char *fnr0;
+    char *fnr1;
+    char *fnr2;
+    char *fn_input; // pointer to input filename in argv do not free
+    bool has12, use_oq, copy_tags;
+    int flag_on, flag_off;
+    fastfile filetype;
+    int def_qual;
+} bam2fq_opts_t;
+
+typedef struct bam2fq_state {
     samFile *fp;
+    FILE *fpse;
+    FILE **fpr;
     bam_hdr_t *h;
-    bam1_t *b;
-    int8_t *buf;
-    int status = EXIT_SUCCESS;
-    size_t max_buf;
-    FILE* fpse;
+    bool has12, use_oq, copy_tags;
+    int flag_on, flag_off;
+    fastfile filetype;
+    int def_qual;
+} bam2fq_state_t;
+
+static readpart which_readpart(const bam1_t *b)
+{
+    if ((b->core.flag & BAM_FREAD1) && !(b->core.flag & BAM_FREAD2)) {
+        return READ_1;
+    } else if ((b->core.flag & BAM_FREAD2) && !(b->core.flag & BAM_FREAD1)) {
+        return READ_2;
+    } else {
+        return READ_UNKNOWN;
+    }
+}
+
+// Transform a bam1_t record into a string with the FASTQ representation of it
+static bool bam1_to_fq(const bam1_t *b, kstring_t *linebuf, const bam2fq_state_t *state)
+{
+    int i;
+    int32_t qlen = b->core.l_qseq;
+    assert(qlen >= 0);
+    uint8_t *seq;
+    uint8_t *qual = bam_get_qual(b);
+    const uint8_t *oq = NULL;
+    if (state->use_oq) oq = bam_aux_get(b, "OQ");
```

Comment:
This was because the +1 in the original code was later on in the code (at the memcpy) and didn't get copied over. Silly me.